### PR TITLE
Add and cleanup RobotState doxygen documentation

### PR DESCRIPTION
**Pull request by davetcoleman**
@jliukkonen asked the difference between getGlobalLinkTransform() and getFrameTransform() so I cleaned up some stuff while verifying my recollection of the difference was correct.

### CODE REVIEW

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h [hunk 1]
```diff
@@ -1311,27 +1298,43 @@ as the new values that correspond to the group */
...
     return getGlobalLinkTransform(robot_model_->getLinkModel(link_name));
   }
 
-  const Eigen::Isometry3d& getGlobalLinkTransform(const LinkModel* link)
-  {
-    updateLinkTransforms();
-    return global_link_transforms_[link->getLinkIndex()];
-  }
-
+  /** \brief Look-up the cached transform from the model frame (root of model) to |link_name|.
```
Comment (mamoll) [inline moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: |link_name| should be |link|

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h [hunk 2]
```diff
@@ -1311,27 +1298,43 @@ as the new values that correspond to the group */
...
   {
     return getCollisionBodyTransform(robot_model_->getLinkModel(link_name), index);
   }
 
-  const Eigen::Isometry3d& getCollisionBodyTransform(const LinkModel* link, std::size_t index)
-  {
-    updateCollisionBodyTransforms();
-    return global_collision_body_transforms_[link->getFirstCollisionBodyTransformIndex() + index];
-  }
+  /** \brief Look-up the cached transform from the model frame (root of model) to |link_name|.
```
Comment (mamoll) [inline moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: |link_name| should be |link|

The |index| parameter is undocumented

**Comment by henningkayser** [comment]
This seems to be overlapping with https://github.com/ros-planning/moveit/pull/1843

**Comment by davetcoleman** [comment]
Accidentally replaces https://github.com/ros-planning/moveit/pull/1843/files

my bad @AndyZe - I've integrated your changes!

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h (hunk 2 above)
Comment (davetcoleman) [reply moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: i believe its:
> specify which collision body to lookup, if more than one exists

**Comment by davetcoleman** [comment]
Please re-review, due to lots of changes

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h [hunk 3]
```diff
@@ -48,6 +48,11 @@
 
 #include <boost/assert.hpp>
 
+/* Terminology
+ *   Model Frame: the root frame of the URDF model
```
Comment (rhaschke) [inline moveit_core/robot_state/include/moveit/robot_state/robot_state.h]:
```suggestion
 *   Model Frame: RobotModel's model frame == PlanningScene's planning frame
     If the SRDF defines a virtual, non-fixed (e.g. floating) joint, this is the parent of this virtual joint.
     Otherwise, it is the root joint of the URDF model.
```
Compare the existing documentation in RobotModel:
https://github.com/ros-planning/moveit/blob/3a57141b48435de0a6026f9fc3496165c72e19e6/moveit_core/robot_model/include/moveit/robot_model/robot_model.h#L455-L457

https://github.com/ros-planning/moveit/blob/3a57141b48435de0a6026f9fc3496165c72e19e6/moveit_core/robot_model/src/robot_model.cpp#L928-L931

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h [hunk 4]
```diff
@@ -1311,27 +1298,39 @@ as the new values that correspond to the group */
   /** \brief Update the state after setting a particular link to the input global transform pose.*/
   void updateStateWithLinkAt(const LinkModel* link, const Eigen::Isometry3d& transform, bool backward = false);
 
-  const Eigen::Isometry3d& getGlobalLinkTransform(const std::string& link_name)
-  {
-    return getGlobalLinkTransform(robot_model_->getLinkModel(link_name));
-  }
+  /** \brief Get the link transform w.r.t. the root link (model frame) of the RobotModel.
+   *   This is typically the root link of the URDF unless a virtual joint is present.
+   *   Checks the cache and if there are any dirty (non-updated) transforms first and updates transforms cache as needed
```
Comment (rhaschke) [inline moveit_core/robot_state/include/moveit/robot_state/robot_state.h]:
```suggestion
   *   Checks the cache and if there are any dirty (non-updated) transforms, first updates them as needed
```

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h [hunk 5]
```diff
@@ -1311,27 +1298,39 @@ as the new values that correspond to the group */
   /** \brief Update the state after setting a particular link to the input global transform pose.*/
   void updateStateWithLinkAt(const LinkModel* link, const Eigen::Isometry3d& transform, bool backward = false);
 
-  const Eigen::Isometry3d& getGlobalLinkTransform(const std::string& link_name)
-  {
-    return getGlobalLinkTransform(robot_model_->getLinkModel(link_name));
```
Comment (rhaschke) [inline moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: Why did you removed the inlining for these 1-liner functions? This (slightly) degrades performance.

Comment (davetcoleman) [reply moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: The optimizer decided what gets inlined, modern compilers do not care if its in the header or not (unless its a template)

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h [hunk 6]
```diff
@@ -1311,27 +1300,39 @@ as the new values that correspond to the group */
...
+  const Eigen::Isometry3d& getGlobalLinkTransform(const LinkModel* link) const;
+
+  /** \brief Get the link transform w.r.t. the root link (model frame) of the RobotModel.
+   *   This is typically the root link of the URDF unless a virtual joint is present.
+   *   Checks the cache and if there are any dirty (non-updated) transforms first and updates transforms cache as needed
+   *   A related, more comprehensive function is |getFrameTransform| which includes this functionality
+   *   but also other frame sources (attached objects, subframes).
+   *
+   *   As opposed to the visual links in |getGlobalLinkTransform|, this function returns
+   *   the collision link transform used for collision checking.
```
Comment (mamoll) [inline moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: This helps explain when to use which one. I find this distinction still a bit confusing. Is this the difference:

- global transform for link: transform for given link in root coordinate frame of robot model
- collision body transform: transform for given collision body in given link's frame

Your high-level description of the transforms returned by the two families methods is the same. I'd reword them a bit to clarify.

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h (hunk 5 above)
Comment (davetcoleman) [reply moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: ``inline`` isn't needed for modern compilers, either. they will choose on their own what to inline or not.

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h (hunk 6 above)
Comment (davetcoleman) [reply moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: They are worded the same because they behave the same. I don't want to spend any more time on this documentation that no one else has bothered to do in the last 8 years. I need to get actual work done.

**Comment by davetcoleman** [comment]
This is ready to be merged.

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h (hunk 5 above)
Comment (rhaschke) [reply moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: The compiler cannot inline this function if it's not declared in the header. How should another compilation unit know what to inline? See http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Rf-inline

**Comment by rhaschke** [comment]
> This is ready to be merged.

I don't agree. I would like to insist on keeping the inlining. 
You are right that the compilers automatically decide when to inline and when not - even with an `inline` statement provided. But they can only do so if the function definition is accessible in the header.

File: moveit_core/robot_state/include/moveit/robot_state/robot_state.h (hunk 5 above)
Comment (davetcoleman) [reply moveit_core/robot_state/include/moveit/robot_state/robot_state.h]: i've added inline to header functions.

**Comment by davetcoleman** [comment]
Okay, I've restored all the functions to be inline in .h file (without inline tag though, which is how it was originally).
